File: backends/java/zappy-domain/src/main/java/com/zappymart/domain/shared/EmailAddress.java
package com.zappymart.domain.shared;

import java.util.Locale;
import java.util.Optional;
import java.util.regex.Pattern;

public record EmailAddress(String value) {

    private static final Pattern SHAPE = Pattern.compile("^[^@\\s]+@[^@\\s.]+(\\.[^@\\s.]+)+$");

    public EmailAddress {
        if (value == null || !SHAPE.matcher(value).matches()) {
            throw new IllegalArgumentException("An email address has a local part, an at sign and a domain");
        }
    }

    public static Optional<EmailAddress> parse(String text) {
        if (text == null) {
            return Optional.empty();
        }
        String normalised = text.trim().toLowerCase(Locale.ROOT);
        if (!SHAPE.matcher(normalised).matches()) {
            return Optional.empty();
        }
        return Optional.of(new EmailAddress(normalised));
    }

    @Override
    public String toString() {
        return value;
    }
}

File: backends/kotlin/zappy-domain/src/main/kotlin/nl/zappymart/domain/shared/EmailAddress.kt
package nl.zappymart.domain.shared

@JvmInline
value class EmailAddress private constructor(val value: String) {

    override fun toString(): String = value

    companion object {
        private val SHAPE = Regex("^[^\\s@]+@[^\\s@]+\\.[^\\s@]+$")

        fun of(text: String): Result<EmailAddress> {
            val normalised = text.trim().lowercase()
            return if (SHAPE.matches(normalised)) {
                Result.Success(EmailAddress(normalised))
            } else {
                refusal(UserErrorCode.EMAIL_INVALID, "That is not an email address.", "input.email")
            }
        }

        fun ofStored(value: String) = EmailAddress(value)
    }
}
